Compiler crash during explicitOuter

The following code crashes Dotty:

```Scala
class A(a: Int) {
  class X {
    val x = a
  }

  trait Y {
    val y = a
  }

  trait Z(val o: A) extends o.Y {
    val z = a
  }

  class B extends X with Z(new A(4))
}
```

<details>
<summary>error message</summary>
<pre>
Exception in thread "main" java.lang.AssertionError: assertion failed: failure to construct path from method A$Y$$$outer/class B/class A/package <empty>/package <root> to `this` of trait Z in class A;
class A does not have an outer accessor
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:36)
	at dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.loop$1(ExplicitOuter.scala:388)
	at dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.path$extension(ExplicitOuter.scala:394)
	at dotty.tools.dotc.transform.LambdaLift$Lifter.memberRef(LambdaLift.scala:400)
	at dotty.tools.dotc.transform.LambdaLift.transformIdent(LambdaLift.scala:519)
	at dotty.tools.dotc.transform.MegaPhase.goIdent(MegaPhase.scala:518)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:210)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:366)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:250)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:367)
	at dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:230)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:233)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:366)
	at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:375)
	at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:380)
	at scala.collection.immutable.List.mapConserve(List.scala:176)
	at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:380)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:319)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:367)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:236)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:366)
	at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:375)
	at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:380)
	at scala.collection.immutable.List.mapConserve(List.scala:176)
	at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:380)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:319)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:367)
	at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:236)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:366)
	at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:375)
	at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:380)
	at scala.collection.immutable.List.mapConserve(List.scala:176)
	at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:380)
	at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:336)
	at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:339)
	at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:367)
	at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:386)
	at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:398)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:295)
	at scala.collection.immutable.List.map(List.scala:283)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:297)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:171)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:186)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:194)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:88)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:201)
	at dotty.tools.dotc.Run.compileSources(Run.scala:136)
	at dotty.tools.dotc.Run.compile(Run.scala:120)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:29)
	at dotty.tools.dotc.Driver.process(Driver.scala:127)
	at dotty.tools.dotc.Driver.process(Driver.scala:96)
	at dotty.tools.dotc.Driver.process(Driver.scala:108)
	at dotty.tools.dotc.Driver.main(Driver.scala:135)
	at dotty.tools.dotc.Main.main(Main.scala)
<pre>
</details>

--- a/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
+++ b/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
@@ -93,7 +93,7 @@ object RefChecks {
   /** Check that self type of this class conforms to self types of parents.
    *  and required classes.
    */
-  private def checkParents(cls: Symbol)(implicit ctx: Context): Unit = cls.info match {
+  private def checkParents(cls: Symbol, tmpl: Template)(implicit ctx: Context): Unit = cls.info match {
     case cinfo: ClassInfo =>
       def checkSelfConforms(other: ClassSymbol, category: String, relation: String) = {
         val otherSelf = other.givenSelfType.asSeenFrom(cls.thisType, other.classSymbol)
@@ -101,12 +101,28 @@ object RefChecks {
           ctx.error(DoesNotConformToSelfType(category, cinfo.selfType, cls, otherSelf, relation, other.classSymbol),
             cls.pos)
       }
-      for (parent <- cinfo.classParents)
-        checkSelfConforms(parent.classSymbol.asClass, "illegal inheritance", "parent")
+      for (parent <- tmpl.parents) {
+        checkSelfConforms(parent.tpe.classSymbol.asClass, "illegal inheritance", "parent")
+        checkParentPrefix(cls, parent)
+      }
       for (reqd <- cinfo.cls.givenSelfType.classSymbols)
         checkSelfConforms(reqd, "missing requirement", "required")
     case _ =>
   }
+
+  /** Disallow using trait parameters as prefix for its parents.
+   *
+   *  The rationale is to ensure outer-related NPE never happen in Scala.
+   *  Otherwise, outer NPE may happen, see tests/neg/i5083.scala
+   */
+  private def checkParentPrefix(cls: Symbol, parent: Tree)(implicit ctx: Context): Unit =
+    parent.tpe.typeConstructor match {
+      case TypeRef(ref: TermRef, _) =>
+        val paramRefs = ref.namedPartsWith(ntp => ntp.symbol.enclosingClass == cls)
+        if (paramRefs.nonEmpty && cls.is(Trait))
+          ctx.error("trait parameters cannot be used as parent prefixes", parent.pos)
+      case _ =>
+    }
 
   /** Check that a class and its companion object to not both define
    *  a class or module with same name
@@ -960,7 +976,7 @@ class RefChecks extends MiniPhase { thisPhase =>
   override def transformTemplate(tree: Template)(implicit ctx: Context) = try {
     val cls = ctx.owner
     checkOverloadedRestrictions(cls)
-    checkParents(cls)
+    checkParents(cls, tree)
     checkCompanionNameClashes(cls)
     checkAllOverrides(cls)
     tree

--- a/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
+++ b/compiler/src/dotty/tools/dotc/typer/RefChecks.scala
@@ -93,7 +93,7 @@ object RefChecks {
   /** Check that self type of this class conforms to self types of parents.
    *  and required classes.
    */
-  private def checkParents(cls: Symbol, tmpl: Template)(implicit ctx: Context): Unit = cls.info match {
+  private def checkParents(cls: Symbol)(implicit ctx: Context): Unit = cls.info match {
     case cinfo: ClassInfo =>
       def checkSelfConforms(other: ClassSymbol, category: String, relation: String) = {
         val otherSelf = other.givenSelfType.asSeenFrom(cls.thisType, other.classSymbol)
@@ -101,10 +101,8 @@ object RefChecks {
           ctx.error(DoesNotConformToSelfType(category, cinfo.selfType, cls, otherSelf, relation, other.classSymbol),
             cls.pos)
       }
-      for (parent <- tmpl.parents) {
-        checkSelfConforms(parent.tpe.classSymbol.asClass, "illegal inheritance", "parent")
-        checkParentPrefix(cls, parent)
-      }
+      for (parent <- cinfo.classParents)
+        checkSelfConforms(parent.classSymbol.asClass, "illegal inheritance", "parent")
       for (reqd <- cinfo.cls.givenSelfType.classSymbols)
         checkSelfConforms(reqd, "missing requirement", "required")
     case _ =>
@@ -119,7 +117,7 @@ object RefChecks {
     parent.tpe.typeConstructor match {
       case TypeRef(ref: TermRef, _) =>
         val paramRefs = ref.namedPartsWith(ntp => ntp.symbol.enclosingClass == cls)
-        if (paramRefs.nonEmpty && cls.is(Trait))
+        if (paramRefs.nonEmpty)
           ctx.error("trait parameters cannot be used as parent prefixes", parent.pos)
       case _ =>
     }
@@ -976,7 +974,8 @@ class RefChecks extends MiniPhase { thisPhase =>
   override def transformTemplate(tree: Template)(implicit ctx: Context) = try {
     val cls = ctx.owner
     checkOverloadedRestrictions(cls)
-    checkParents(cls, tree)
+    checkParents(cls)
+    if (cls.is(Trait)) tree.parents.foreach(checkParentPrefix(cls, _))
     checkCompanionNameClashes(cls)
     checkAllOverrides(cls)
     tree

--- a/tests/neg/i5083.scala
+++ b/tests/neg/i5083.scala
@@ -1,0 +1,25 @@
+class A(a: Int) {
+  abstract class X {
+    def f: Int
+    val x = a + f      // NPE: the outer for `Y` is not yet set
+  }
+
+  trait Y {
+    val y = a
+    def f: Int = A.this.a      // NPE: the outer for `Y` is not yet set
+  }
+
+  trait Z(val o: A) extends o.Y {  // error
+    val z = a
+  }
+
+  class B extends X with Z(new A(4))
+}
+
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    val a = new A(3)
+    new a.B
+  }
+}

--- a/tests/neg/i5083b.scala
+++ b/tests/neg/i5083b.scala
@@ -1,0 +1,15 @@
+class A(a: Int) {
+  class X {
+    val x = a
+  }
+
+  trait Y {
+    val y = a
+  }
+
+  val z: Z = ???
+
+  trait Z(val o: A) extends z.o.Y   // error: cyclic reference `z`
+
+  class B extends X with Z(new A(4))
+}

--- a/tests/neg/i5083c.scala
+++ b/tests/neg/i5083c.scala
@@ -1,0 +1,25 @@
+class A(a: Int) {
+  trait X {
+    def f: Int
+    val x = a + f      // NPE: the outer for `Y` is not yet set
+  }
+
+  trait Y {
+    val y = a
+    def f: Int = A.this.a      // NPE: the outer for `Y` is not yet set
+  }
+
+  class Z(o: A) extends m.Y {  // error
+    val m = o
+  }
+
+  class B extends Z(new A(4))
+}
+
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    val a = new A(3)
+    new a.B
+  }
+}

